db.js

```diff
@@ -1,192 +1,261 @@
+    if (url && typeof url === "object") {
+      return url;
     }
+
+    let response;
+    // Try fetching directly for all URLs (including P2P URLs)
     // TODO: Signed fetch
-    const response = await this.fetch.call(globalThis, url, {
-      headers: {
-        Accept: 'application/json'
+    try {
+      response = await this.fetch.call(globalThis, url, {
+        headers: {
+          Accept: "application/json",
+        },
+      });
+    } catch (error) {
+      console.error("P2P loading failed, trying HTTP gateway:", error);
+    }
+
+    // If direct fetch was not successful, attempt fetching from a gateway for P2P protocols
+    if (!response || !response.ok) {
+      let gatewayUrl = url;
+
+      if (url.startsWith("hyper://")) {
+        gatewayUrl = url.replace("hyper://", "https://hyper.hypha.coop/hyper/");
+      } else if (url.startsWith("ipns://")) {
+        gatewayUrl = url.replace("ipns://", "https://ipfs.hypha.coop/ipns/");
+      }
+
+      try {
+        response = await this.fetch.call(globalThis, gatewayUrl, {
+          headers: {
+            Accept: "application/json",
+          },
+        });
+      } catch (error) {
+        console.error("Fetching from gateway failed:", error);
+        throw new Error(`Failed to fetch ${url} from gateway`);
       }
-    })
-    if (!response.ok) {
-      throw new Error(await response.text())
     }
 
-    return await response.json()
+    if (!response.ok) {
+      throw new Error(`HTTP error! Status: ${response.status}`);
+    }
+    return await response.json();
   }
 
-  async getActor (url) {
+  async getActor(url) {
     // TODO: Try to load from cache
-    const actor = await this.#get(url)
-    this.db.put(ACTORS_STORE, actor)
-    return this.db.get(ACTORS_STORE, actor.id)
+    const actor = await this.#get(url);
+    this.db.put(ACTORS_STORE, actor);
+    return this.db.get(ACTORS_STORE, actor.id);
+  }
+
+  async getAllActors() {
+    const tx = this.db.transaction(ACTORS_STORE);
+    const actors = [];
+    let cursor = await tx.store.openCursor();
+
+    while (cursor) {
+      actors.push(cursor.value);
+      cursor = await cursor.continue();
+    }
+
+    return actors;
   }
 
-  async getNote (url) {
+  async getNote(url) {
     try {
-      return this.db.get(NOTES_STORE, url)
+      return this.db.get(NOTES_STORE, url);
     } catch {
-      const note = await this.#get(url)
-      await this.ingestNote(note)
-      return note
+      const note = await this.#get(url);
+      await this.ingestNote(note);
+      return note;
     }
   }
 
-  async * searchNotes (query) {
-    // Query can search by any of the indexed fields.
-    // Everything gets sorted by the `published` time.
-    // We should find a way to allow for arrays of values and open multiple iterators
-    // Kinda like this: https://git.sr.ht/~breatheoutbreathein/ushin-db/tree/master/item/USHINBase.js#L509
-
-    const tx = this.db.transaction(NOTES_STORE)
+  async ingestActor(url) {
```

Comment:
Are all these changes necessary? Feels like we should move the logic for determening if the actor.outbox is a collection into `iterateCollection`